flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotRequestBulkTracker.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.jobmaster.slotpool;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.jobmaster.SlotInfo;
+import org.apache.flink.util.clock.Clock;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Tracks physical slot request bulks. Once a bulk is registered, a fulfill-ability check for it will be scheduled.
+ */
+class PhysicalSlotRequestBulkTracker {
+
+	private final Supplier<Set<SlotInfo>> slotsRetriever;
+
+	private final Clock clock;
+
+	private final Set<PhysicalSlotRequestBulk> slotRequestBulks;
```

Comment:
Not sure I understand why we need `slotRequestBulks/track/untrack/isTracked`.
Looks only for tests. Can we remove it?

Reply:
Yes we can remove them. I had wanted to keep the bulks tracked for sanity check and diagnostic but of course it is not necessary.

Reply:
Done via af7d4926ee8bb041713feeed10770147509d2db8.
`PhysicalSlotRequestBulkTracker#createPhysicalSlotRequestBulk` is introduced so that we can reset the unfulfillable timestamp for a newly created bulk with the clock of `PhysicalSlotRequestBulkTracker`.

Reply:
What do you think of rename `PhysicalSlotRequestBulkTracker` to `PhysicalSlotRequestBulkHelper` since it does not track the bulks but provides several methods which helps with bulk creation, initialization and checking?

Reply:
Maybe then `PhysicalSlotRequestBulkChecker`?

Reply:
`PhysicalSlotRequestBulkChecker` sounds good to me.

Reply:
done.